From cd6e830faaa9a29227f7ab85509342a758f1cef7 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sun, 26 Oct 2025 18:02:08 +0100
Subject: [PATCH 0998/1013] evl/net: udp: use source address obtained from
 routing

On transmit in unconnected mode (i.e. no address was bound to the
socket), use the source address determined when the routing
information was generated then passed to evl_net_learn_ipv4_route().

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 kernel/evl/net/ipv4/udp.c | 16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

diff --git a/kernel/evl/net/ipv4/udp.c b/kernel/evl/net/ipv4/udp.c
index fea590648249..3917fd71929a 100644
--- a/kernel/evl/net/ipv4/udp.c
+++ b/kernel/evl/net/ipv4/udp.c
@@ -443,15 +443,19 @@ static ssize_t send_udp(struct evl_socket *esk,
 		return -EINPROGRESS;
 	}
 
-	/* Ok, we have an oob path for that datagram. */
-
+	/*
+	 * Ok, we have an oob path for that datagram. In connected
+	 * mode (i.e. an address was bound to the socket), pick the
+	 * bound source address. Otherwise, use the source address
+	 * determined when the routing information was established
+	 * then passed to evl_net_learn_ipv4_route().
+	 */
 	saddr = inet->inet_saddr;
 	if (!saddr) {
-		if (ipv4_is_multicast(daddr)) {
+		if (ipv4_is_multicast(daddr))
 			saddr = inet->mc_addr;
-		} else {
-			saddr = inet_select_addr(ert->rt->dst.dev, daddr, RT_SCOPE_LINK);
-		}
+		else
+			saddr = ert->flowi4.saddr;
 	}
 
 	ipc.saddr = saddr;
-- 
2.30.2

